File: python/toolsets/event-correlator/src/src/event_correlator/postgres.py
from __future__ import annotations

from collections.abc import Callable
from datetime import datetime
from importlib.resources import files
from typing import Any

import psycopg
from psycopg.rows import dict_row
from psycopg.types.json import Jsonb

from .models import PendingTrade

AuditBuilder = Callable[[list[PendingTrade]], dict[str, Any]]
BreakAuditBuilder = Callable[[str, list[PendingTrade]], dict[str, Any]]


class PostgresStore:
    def __init__(self, database_url: str) -> None:
        self._connection = psycopg.connect(
            database_url,
            autocommit=True,
            row_factory=dict_row,
            connect_timeout=5,
        )

    def close(self) -> None:
        self._connection.close()

    def migrate(self) -> None:
        migration = (
            files("event_correlator.migrations")
            .joinpath("00001_initial_schema.sql")
            .read_text(encoding="utf-8")
        )
        self._connection.execute(migration)

    def ingest(
        self,
        pending: PendingTrade,
        expected_sources: list[str],
        build_audit: AuditBuilder,
    ) -> tuple[bool, list[PendingTrade]]:
        with self._connection.transaction():
            self._connection.execute(
                "SELECT pg_advisory_xact_lock(hashtextextended(%s, 0))",
                (pending.trade_id,),
            )
            finalized = self._connection.execute(
                "SELECT EXISTS (SELECT 1 FROM audit_log WHERE trade_id = %s) AS finalized",
                (pending.trade_id,),
            ).fetchone()
            if finalized and finalized["finalized"]:
                return False, []

            existing = self._get_events(pending.trade_id)
            if existing:
                pending = PendingTrade(
                    trade_id=pending.trade_id,
                    source=pending.source,
                    payload=pending.payload,
                    first_seen=min(event.first_seen for event in existing),
                    deadline=min(event.deadline for event in existing),
                )

            inserted = self._connection.execute(
                """
                INSERT INTO pending_events
                    (trade_id, source, payload, first_seen, deadline)
                VALUES (%s, %s, %s, %s, %s)
                ON CONFLICT (trade_id, source) DO NOTHING
                RETURNING trade_id
                """,
                (
                    pending.trade_id,
                    pending.source,
                    Jsonb(pending.payload),
                    pending.first_seen,
                    pending.deadline,
                ),
            ).fetchone()
            if inserted is None:
                return False, []

            events = self._get_events(pending.trade_id)
            received_sources = {event.source for event in events}
            if not set(expected_sources).issubset(received_sources):
                return True, []

            expected_events = [
                event for event in events if event.source in set(expected_sources)
            ]
            self._connection.execute(
                "DELETE FROM pending_events WHERE trade_id = %s",
                (pending.trade_id,),
            )
            self._record_audit(build_audit(expected_events))
            return True, expected_events

    def sweep_expired(
        self,
        now: datetime,
        build_audit: BreakAuditBuilder,
    ) -> dict[str, list[PendingTrade]]:
        rows = self._connection.execute(
            """
            SELECT DISTINCT trade_id
            FROM pending_events
            WHERE deadline <= %s
            ORDER BY trade_id
            """,
            (now,),
        ).fetchall()

        expired: dict[str, list[PendingTrade]] = {}
        for row in rows:
            trade_id = row["trade_id"]
            with self._connection.transaction():
                lock = self._connection.execute(
                    "SELECT pg_try_advisory_xact_lock(hashtextextended(%s, 0)) AS locked",
                    (trade_id,),
                ).fetchone()
                if not lock or not lock["locked"]:
                    continue

                events = self._get_events(trade_id)
                if not events or min(event.deadline for event in events) > now:
                    continue

                self._connection.execute(
                    "DELETE FROM pending_events WHERE trade_id = %s",
                    (trade_id,),
                )
                self._record_audit(build_audit(trade_id, events))
                expired[trade_id] = events
        return expired

    def get_audit(
        self,
        trade_id: str | None,
        outcome: str | None,
        limit: int,
    ) -> list[dict[str, Any]]:
        rows = self._connection.execute(
            """
            SELECT id::text, trade_id, outcome, sources, detail,
                   occurred_at, recorded_at
            FROM audit_log
            WHERE (%s::text IS NULL OR trade_id = %s::text)
              AND (%s::text IS NULL OR outcome = %s::text)
            ORDER BY recorded_at DESC, id DESC
            LIMIT %s
            """,
            (trade_id, trade_id, outcome, outcome, min(max(limit, 1), 500)),
        ).fetchall()
        return [_json_ready(dict(row)) for row in rows]

    def _get_events(self, trade_id: str) -> list[PendingTrade]:
        rows = self._connection.execute(
            """
            SELECT trade_id, source, payload, first_seen, deadline
            FROM pending_events
            WHERE trade_id = %s
            ORDER BY source
            FOR UPDATE
            """,
            (trade_id,),
        ).fetchall()
        return [
            PendingTrade(
                trade_id=row["trade_id"],
                source=row["source"],
                payload=row["payload"],
                first_seen=row["first_seen"],
                deadline=row["deadline"],
            )
            for row in rows
        ]

    def _record_audit(self, entry: dict[str, Any]) -> None:
        self._connection.execute(
            """
            INSERT INTO audit_log (trade_id, outcome, sources, detail, occurred_at)
            VALUES (%s, %s, %s, %s, %s)
            """,
            (
                entry["trade_id"],
                entry["outcome"],
                Jsonb(entry["sources"]),
                Jsonb(entry["detail"]),
                entry["occurred_at"],
            ),
        )


def _json_ready(value: Any) -> Any:
    if isinstance(value, datetime):
        return value.isoformat().replace("+00:00", "Z")
    if isinstance(value, dict):
        return {key: _json_ready(item) for key, item in value.items()}
    if isinstance(value, list):
        return [_json_ready(item) for item in value]
    return value

File: go/toolsets/event-correlator/src/internal/store/postgres/postgres.go
// Package postgres implements the correlator store backed by PostgreSQL.
package postgres

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/stdlib"

	"github.com/solacese/event-correlator-tool/go/toolsets/event-correlator/src/internal/model"
	"github.com/solacese/event-correlator-tool/go/toolsets/event-correlator/src/internal/store"
)

// Store implements store.Store backed by PostgreSQL.
type Store struct {
	db *sql.DB
}

// New opens a connection pool to the given Postgres DSN.
func New(databaseURL string) (*Store, error) {
	connConfig, err := pgx.ParseConfig(databaseURL)
	if err != nil {
		return nil, fmt.Errorf("parse database URL: %w", err)
	}
	db := stdlib.OpenDB(*connConfig)
	db.SetMaxOpenConns(5)
	db.SetMaxIdleConns(2)
	db.SetConnMaxLifetime(5 * time.Minute)

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := db.PingContext(ctx); err != nil {
		_ = db.Close()
		return nil, fmt.Errorf("ping database: %w", err)
	}
	return &Store{db: db}, nil
}

func (s *Store) Migrate(ctx context.Context) error {
	migration, err := migrationSQL()
	if err != nil {
		return fmt.Errorf("read migration: %w", err)
	}
	if _, err := s.db.ExecContext(ctx, migration); err != nil {
		return fmt.Errorf("run migration: %w", err)
	}
	return nil
}

func (s *Store) Ingest(ctx context.Context, trade store.PendingTrade, expectedSources []string, buildAudit func([]store.PendingTrade) model.AuditEntry) (result store.IngestResult, err error) {
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return result, fmt.Errorf("begin transaction: %w", err)
	}
	defer rollback(tx)

	if _, err = tx.ExecContext(ctx, `SELECT pg_advisory_xact_lock(hashtextextended($1, 0))`, trade.TradeID); err != nil {
		return result, fmt.Errorf("lock trade: %w", err)
	}

	var finalized bool
	if err := tx.QueryRowContext(ctx, `SELECT EXISTS (SELECT 1 FROM audit_log WHERE trade_id = $1)`, trade.TradeID).Scan(&finalized); err != nil {
		return result, fmt.Errorf("check finalized trade: %w", err)
	}
	if finalized {
		if err := tx.Commit(); err != nil {
			return result, fmt.Errorf("commit finalized check: %w", err)
		}
		return result, nil
	}

	existing, err := getEvents(ctx, tx, trade.TradeID)
	if err != nil {
		return result, err
	}
	for _, event := range existing {
		if event.Deadline.Before(trade.Deadline) {
			trade.Deadline = event.Deadline
		}
	}

	insert, err := tx.ExecContext(ctx, `
		INSERT INTO pending_events (trade_id, source, payload, first_seen, deadline)
		VALUES ($1, $2, $3, $4, $5)
		ON CONFLICT (trade_id, source) DO NOTHING`,
		trade.TradeID, trade.Source, trade.Payload, trade.FirstSeen, trade.Deadline,
	)
	if err != nil {
		return result, fmt.Errorf("record event: %w", err)
	}
	rowsAffected, err := insert.RowsAffected()
	if err != nil {
		return result, fmt.Errorf("read insert result: %w", err)
	}
	if rowsAffected == 0 {
		if err := tx.Commit(); err != nil {
			return result, fmt.Errorf("commit duplicate: %w", err)
		}
		return result, nil
	}
	result.Inserted = true

	events, err := getEvents(ctx, tx, trade.TradeID)
	if err != nil {
		return result, err
	}
	if !hasEverySource(events, expectedSources) {
		if err := tx.Commit(); err != nil {
			return result, fmt.Errorf("commit pending event: %w", err)
		}
		return result, nil
	}

	result.Events = expectedEvents(events, expectedSources)
	if _, err := tx.ExecContext(ctx, `DELETE FROM pending_events WHERE trade_id = $1`, trade.TradeID); err != nil {
		return result, fmt.Errorf("delete reconciled trade: %w", err)
	}
	if err := recordAudit(ctx, tx, buildAudit(result.Events)); err != nil {
		return result, err
	}
	if err := tx.Commit(); err != nil {
		return result, fmt.Errorf("commit reconciliation: %w", err)
	}
	return result, nil
}

func (s *Store) SweepExpired(ctx context.Context, now time.Time, buildAudit func(string, []store.PendingTrade) model.AuditEntry) (map[string][]store.PendingTrade, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT DISTINCT trade_id
		FROM pending_events
		WHERE deadline <= $1
		ORDER BY trade_id`, now)
	if err != nil {
		return nil, fmt.Errorf("get expired trades: %w", err)
	}
	var tradeIDs []string
	for rows.Next() {
		var tradeID string
		if err := rows.Scan(&tradeID); err != nil {
			rows.Close()
			return nil, fmt.Errorf("scan expired trade: %w", err)
		}
		tradeIDs = append(tradeIDs, tradeID)
	}
	if err := rows.Close(); err != nil {
		return nil, fmt.Errorf("close expired rows: %w", err)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate expired trades: %w", err)
	}

	expired := make(map[string][]store.PendingTrade)
	for _, tradeID := range tradeIDs {
		events, claimed, err := s.claimExpired(ctx, tradeID, now, buildAudit)
		if err != nil {
			return nil, err
		}
		if claimed {
			expired[tradeID] = events
		}
	}
	return expired, nil
}

func (s *Store) claimExpired(ctx context.Context, tradeID string, now time.Time, buildAudit func(string, []store.PendingTrade) model.AuditEntry) (events []store.PendingTrade, claimed bool, err error) {
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return nil, false, fmt.Errorf("begin sweep transaction: %w", err)
	}
	defer rollback(tx)

	var locked bool
	if err := tx.QueryRowContext(ctx, `SELECT pg_try_advisory_xact_lock(hashtextextended($1, 0))`, tradeID).Scan(&locked); err != nil {
		return nil, false, fmt.Errorf("lock expired trade %q: %w", tradeID, err)
	}
	if !locked {
		return nil, false, nil
	}

	events, err = getEvents(ctx, tx, tradeID)
	if err != nil {
		return nil, false, err
	}
	if len(events) == 0 || earliestDeadline(events).After(now) {
		if err := tx.Commit(); err != nil {
			return nil, false, fmt.Errorf("commit skipped sweep: %w", err)
		}
		return nil, false, nil
	}

	if _, err := tx.ExecContext(ctx, `DELETE FROM pending_events WHERE trade_id = $1`, tradeID); err != nil {
		return nil, false, fmt.Errorf("delete expired trade: %w", err)
	}
	if err := recordAudit(ctx, tx, buildAudit(tradeID, events)); err != nil {
		return nil, false, err
	}
	if err := tx.Commit(); err != nil {
		return nil, false, fmt.Errorf("commit break: %w", err)
	}
	return events, true, nil
}

func (s *Store) GetAudit(ctx context.Context, filter store.AuditFilter) ([]model.AuditEntry, error) {
	limit := filter.Limit
	if limit <= 0 {
		limit = 100
	}
	if limit > 500 {
		limit = 500
	}
	rows, err := s.db.QueryContext(ctx, `
		SELECT id::text, trade_id, outcome, sources, detail, occurred_at, recorded_at
		FROM audit_log
		WHERE ($1 = '' OR trade_id = $1)
		  AND ($2 = '' OR outcome = $2)
		ORDER BY recorded_at DESC, id DESC
		LIMIT $3`, filter.TradeID, filter.Outcome, limit)
	if err != nil {
		return nil, fmt.Errorf("query audit history: %w", err)
	}
	defer rows.Close()

	entries := make([]model.AuditEntry, 0)
	for rows.Next() {
		var entry model.AuditEntry
		var sourcesJSON []byte
		if err := rows.Scan(&entry.ID, &entry.TradeID, &entry.Outcome, &sourcesJSON, &entry.Detail, &entry.OccurredAt, &entry.RecordedAt); err != nil {
			return nil, fmt.Errorf("scan audit history: %w", err)
		}
		if err := json.Unmarshal(sourcesJSON, &entry.Sources); err != nil {
			return nil, fmt.Errorf("decode audit sources: %w", err)
		}
		entries = append(entries, entry)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate audit history: %w", err)
	}
	return entries, nil
}

func (s *Store) Close() error {
	return s.db.Close()
}

func getEvents(ctx context.Context, tx *sql.Tx, tradeID string) ([]store.PendingTrade, error) {
	rows, err := tx.QueryContext(ctx, `
		SELECT trade_id, source, payload, first_seen, deadline
		FROM pending_events
		WHERE trade_id = $1
		ORDER BY source
		FOR UPDATE`, tradeID)
	if err != nil {
		return nil, fmt.Errorf("get events: %w", err)
	}
	defer rows.Close()

	events := make([]store.PendingTrade, 0)
	for rows.Next() {
		var event store.PendingTrade
		if err := rows.Scan(&event.TradeID, &event.Source, &event.Payload, &event.FirstSeen, &event.Deadline); err != nil {
			return nil, fmt.Errorf("scan event: %w", err)
		}
		events = append(events, event)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate events: %w", err)
	}
	return events, nil
}

func recordAudit(ctx context.Context, tx *sql.Tx, entry model.AuditEntry) error {
	sourcesJSON, err := json.Marshal(entry.Sources)
	if err != nil {
		return fmt.Errorf("marshal audit sources: %w", err)
	}
	if _, err := tx.ExecContext(ctx, `
		INSERT INTO audit_log (trade_id, outcome, sources, detail, occurred_at)
		VALUES ($1, $2, $3, $4, $5)`,
		entry.TradeID, entry.Outcome, sourcesJSON, entry.Detail, entry.OccurredAt,
	); err != nil {
		return fmt.Errorf("record audit: %w", err)
	}
	return nil
}

func rollback(tx *sql.Tx) {
	_ = tx.Rollback()
}

func hasEverySource(events []store.PendingTrade, expectedSources []string) bool {
	seen := make(map[string]struct{}, len(events))
	for _, event := range events {
		seen[event.Source] = struct{}{}
	}
	for _, source := range expectedSources {
		if _, ok := seen[source]; !ok {
			return false
		}
	}
	return true
}

func expectedEvents(events []store.PendingTrade, expectedSources []string) []store.PendingTrade {
	expected := make(map[string]struct{}, len(expectedSources))
	for _, source := range expectedSources {
		expected[source] = struct{}{}
	}
	filtered := make([]store.PendingTrade, 0, len(expectedSources))
	for _, event := range events {
		if _, ok := expected[event.Source]; ok {
			filtered = append(filtered, event)
		}
	}
	return filtered
}

func earliestDeadline(events []store.PendingTrade) time.Time {
	earliest := events[0].Deadline
	for _, event := range events[1:] {
		if event.Deadline.Before(earliest) {
			earliest = event.Deadline
		}
	}
	return earliest
}
